src/lib/utils/plugin_manager.cpp

```diff
@@ -60,21 +66,48 @@ void PluginManager::load_plugin(const std::filesystem::path& path) {
          "Loading plugin failed: There can only be one instance of every plugin.");
 
   plugin_handle_wrapper.plugin->start();
-  _plugins[name] = std::move(plugin_handle_wrapper);
+  _plugins[plugin_name] = std::move(plugin_handle_wrapper);
+
+  // Add the newly loaded plugin's user executable functions to our map of functions.
+  const auto user_executable_functions = _plugins[plugin_name].plugin->provided_user_executable_functions();
+  for (const auto& [function_name, function_pointer] : user_executable_functions) {
+    _user_executable_functions[{plugin_name, function_name}] = function_pointer;
+  }
 }
 
-void PluginManager::unload_plugin(const PluginName& name) {
-  auto plugin_iter = _plugins.find(name);
-  Assert(plugin_iter != _plugins.cend(), "Unloading plugin failed: A plugin with name " + name + " does not exist.");
+void PluginManager::exec_user_function(const PluginName& plugin_name, const PluginFunctionName& function_name) {
+  Assert(_user_executable_functions.count({plugin_name, function_name}) > 0,
+         "There is no " + function_name + " defined for plugin " + plugin_name + ".");
+
+  const auto user_executable_function = _user_executable_functions[{plugin_name, function_name}];
+  user_executable_function();
+
+  auto& log_manager = Hyrise::get().log_manager;
+  log_manager.add_message(
+      "PluginManager",
+      "Called user executable function `" + function_name + "` provided by plugin `" + plugin_name + "`.",
```

Comment:
First to the general discussion on how to handle long running functions. I talked to @Bouncner about that. We would leave it to the plugin to take care of long running functions and document that, e.g., in the AbstractPlugin.

Making users aware is a good idea. However, users might get a distorted picture if the blocking call of the PluginManager terminates quickly but the actual functionality executed by the plugin itself in a separate thread runs forever. I don't have a strong opinion here but I would not add wall clock information. Maybe we can get a third opinion @Bouncner?

Reply:
Hmm, I think extending the message by "(function returned after 17 ms)" can't hurt and might help in certain situations.
In case somebody thinks starting a thread in the function is a good idea, this person should be aware that one cannot rely on the runtime in the log message. What users are we talking about, when we talk about "users might get a distorted pictures"? I'd assume we'll never have the situation where the "user" executing the plugin has not been (at partially) involved in developing the plugin. I wouldn't see a problem in this case.

Or did I get you wrong?

Reply:
I agree that it won't hurt. On the other hand, this is a user-triggered blocking function. Hence, if the function takes forever, the user will notice because the function will not return quickly. If the function takes 17ms the call returns immediately. On the console or via a client, the response typically even contains the execution time.

No, I think you understood me correctly. Plugin users will probably be involved in plugin development. However, plugins might be released as part of research papers where the user was not involved in the development. 

All in all, it's easy to add the timing information and it does not come with much overhead. However, I don't see much benefit of it. In particular because we want the PluginManager's blocking call to always finish really quickly, because the user is usually involved in the development, and we want the heavy lifting to be done inside the plugin itself.